source/slang/slang-ir-link.cpp

```diff
@@ -1182,18 +1189,23 @@ IRInst* cloneGlobalValueWithLinkage(
 
     // We will try to track the "best" declaration we can find.
     //
-    // Generally, one declaration wil lbe better than another if it is
+    // Generally, one declaration will be better than another if it is
     // more specialized for the chosen target. Otherwise, we simply favor
     // definitions over declarations.
     //
-    IRInst* bestVal = sym->irGlobalValue;
-    for( auto ss = sym->nextWithSameName; ss; ss = ss->nextWithSameName )
+    IRInst* bestVal = nullptr;
+    for(IRSpecSymbol* ss = sym; ss; ss = ss->nextWithSameName )
```

Comment:
Is there a problem that came up with the way the code was phrased before, or is this just a stylistic refactor?

It seems like this code is trying to be defensive against an `IRSpecSymbol` where the `irGlobalValue` is null, but that case shouldn't be happening. Was that something that was coming up during debugging?

Reply:
In the end the result is the same, but when the change was made the thought was that the first entry may not be 'best' - for example an ExternDeclaration when that was there was never best, and so not set. As it stands the isBetterForTarget just accepts anything if it has nothing now, so the behavior is the same as before.

Reply:
Hmm... okay, it sounds like it was an intermediate step during implementation of the feature where the `irGlobalValue` field might actually be null.

> was never best

The linker logic is intentionally supposed to be set up so that any IR value with the given mangled name could be the "best" for a given linker invocation, and there is no category of IR value/symbol that we can say a prior is "never best." In particular, if all of the IR values had `[import(...)]` decorations -- which includes the case where there is only one IR value with the given name -- then it is possible for the linker to select a non-definition value as the "best." This is an intentional feature because it means that if you have modules A, B, and C where A and B both depend on C, then it is safe to link A and B together (without C) and you'd still end up with a usable module, albeit one that still has `[import(...)]` decorations for the things it wants to import from C.